File: windows/HarnessPortable.Windows/Services/TunnelManager.cs
namespace HarnessPortable.Windows.Services;

/// <summary>
/// Owns one <see cref="TunnelEngine"/> per tunnel profile so the desktop app
/// can run several SSH tunnels at the same time. Each engine keeps its own
/// state hub; this class re-publishes their events in a single stream.
/// </summary>
public sealed class TunnelManager
{
    private readonly ProfileStore _profiles;
    private readonly SecureStore _secrets;
    private readonly KnownHostsStore _knownHosts;
    private readonly object _gate = new();
    private readonly Dictionary<string, TunnelEngine> _engines = [];

    public TunnelManager(ProfileStore profiles, SecureStore secrets, KnownHostsStore knownHosts)
    {
        _profiles = profiles;
        _secrets = secrets;
        _knownHosts = knownHosts;
    }

    public event Action<TunnelInfo>? StateChanged;

    public void Start(string profileId)
    {
        var engine = GetOrCreate(profileId);
        engine.Start(profileId);
    }

    public void Stop(string profileId, bool announce = true)
    {
        TunnelEngine? engine;
        lock (_gate)
        {
            _engines.TryGetValue(profileId, out engine);
        }

        engine?.Stop(announce);
    }

    public void StopAll(bool announce = false)
    {
        TunnelEngine[] engines;
        lock (_gate)
        {
            engines = _engines.Values.ToArray();
        }

        foreach (var engine in engines)
        {
            engine.Stop(announce);
        }
    }

    public TunnelInfo GetState(string profileId)
    {
        TunnelEngine? engine;
        lock (_gate)
        {
            _engines.TryGetValue(profileId, out engine);
        }

        return engine?.State.Current ?? new TunnelInfo(profileId, Status: TunnelStatus.Stopped);
    }

    public IEnumerable<TunnelInfo> GetActiveStates()
    {
        TunnelEngine[] engines;
        lock (_gate)
        {
            engines = _engines.Values.ToArray();
        }

        foreach (var engine in engines)
        {
            var state = engine.State.Current;
            if (state.Status is TunnelStatus.Connecting or TunnelStatus.Connected or TunnelStatus.Retrying)
            {
                yield return state;
            }
        }
    }

    private TunnelEngine GetOrCreate(string profileId)
    {
        lock (_gate)
        {
            if (_engines.TryGetValue(profileId, out var existing))
            {
                return existing;
            }

            var engine = new TunnelEngine(_profiles, _secrets, _knownHosts);
            engine.State.StateChanged += info => StateChanged?.Invoke(info);
            _engines[profileId] = engine;
            return engine;
        }
    }
}

File: macos/HarnessPortable/Services/TunnelManager.swift
import Foundation
import Combine

@MainActor
final class TunnelManager: ObservableObject {
    @Published private(set) var states: [String: TunnelInfo] = [:]

    private let keychain: KeychainStore
    private let knownHosts: KnownHostsStore
    private var engines: [String: SSHProcessTunnel] = [:]
    private var profiles: [String: TunnelProfile] = [:]
    private var isShuttingDown = false

    init(keychain: KeychainStore, knownHosts: KnownHostsStore) {
        self.keychain = keychain
        self.knownHosts = knownHosts
    }

    func start(_ profile: TunnelProfile) {
        guard !isShuttingDown else { return }
        profiles[profile.id] = profile
        let current = states[profile.id]?.status
        if current == .connected || current == .connecting || current == .retrying { return }
        let engine = engine(for: profile.id)
        let generation = engine.start(profile: profile)
        states[profile.id] = TunnelInfo(
            profileID: profile.id,
            profileName: profile.displayName,
            status: .connecting,
            message: "正在连接…",
            generation: generation
        )
    }

    func stop(_ profileID: String, announce: Bool = true) {
        guard let engine = engines[profileID], let profile = profiles[profileID] else { return }
        let generation = engine.stop(profile: profile)
        states[profileID] = TunnelInfo(
            profileID: profileID,
            profileName: profile.displayName,
            status: .stopped,
            message: announce ? "隧道已停止" : nil,
            generation: generation
        )
    }

    func stopAll() {
        let profileIDs = Array(engines.keys)
        for profileID in profileIDs {
            stop(profileID, announce: false)
        }
    }

    func shutdown() {
        guard !isShuttingDown else { return }
        isShuttingDown = true
        let currentEngines = engines
        for (profileID, engine) in currentEngines {
            if let profile = profiles[profileID] {
                engine.shutdown(profile: profile)
            } else {
                engine.shutdown()
            }
        }
    }

    func state(for profileID: String) -> TunnelInfo {
        states[profileID] ?? TunnelInfo(profileID: profileID, status: .stopped)
    }

    var activeStates: [TunnelInfo] {
        states.values.filter { info in
            info.status == .connecting || info.status == .connected || info.status == .retrying
        }.sorted { ($0.profileName ?? "") < ($1.profileName ?? "") }
    }

    private func engine(for profileID: String) -> SSHProcessTunnel {
        if let existing = engines[profileID] { return existing }
        let engine = SSHProcessTunnel(keychain: keychain, knownHosts: knownHosts)
        engine.onState = { [weak self] info, generation in
            DispatchQueue.main.async { [weak self] in
                guard let self,
                      !self.isShuttingDown,
                      self.engines[profileID]?.isCurrent(generation) == true else { return }
                self.states[profileID] = info
            }
        }
        engines[profileID] = engine
        return engine
    }
}
